src/middleware.ts

```diff
@@ -0,0 +1,21 @@
+import { NextRequest, NextResponse } from "next/server";
+
+/**
+ * Our API routes are already protected, this middleware applies to UI routes like /students.
+ * If a user isn't logged in, we redirect them to sign-in.
+ * This **does not** check for the validity of the session, only that a cookie exists; so it shouldn't be generally used to protect routes.
+ *
+ * Ideally we would use NextAuth's built-in Next.js middleware: https://next-auth.js.org/configuration/nextjs#middleware
+ * However, their middleware doesn't yet work with database sessions.
+ */
+export function middleware(request: NextRequest) {
+  if (!request.cookies.get("next-auth.session-token")) {
+    const url = request.nextUrl.clone();
```

Comment:
How you put these properties together really shows the thoughtfulness you use when coding.